File: Languages/Julia/Go/ccall/invoke.jl
# Memanggil kode Go dari Julia (JVM)
#
# Ditulis untuk Julia v0.6+
# https://github.com/JuliaLang/julia
#
# Run:
#   $ julia invoke.jl 


# Beberapa tipe di Go dapat dipetakan secara langsung sebagai tipe primitif di C++, 
# sebagian lagi dapat dianggap seagai tipe majemuk / struktur.

# Umumnya ketika membuat DLL dengan Go, compiler akan membuat sebuah file header
# untuk memudahkan interfacing. Namun dalam kasus tertentu kita hanya mendapatkan file
# DLL tanpa adanya header. Untuk melakukan interfacing, kita perlu mendefinisikan sendiri
# fungsi-fungsi yang diekspor oleh DLL.

##### Definisikan tipe #####
struct GoSlice
    arr::Ptr{Void}
    len::Int64
    cap::Int64
end
GoSlice(a::Vector, cap=10) = GoSlice(pointer(a), length(a), cap)

struct GoStr
    p::Ptr{Cchar}
    len::Int64
end
GoStr(s::String) = GoStr(pointer(s), length(s))


##### Load library #####
const called = "called.dll"


##### Deskripsikan signature setiap fungsi #####
Add(x,y)   = ccall((:Add, called), Int,(Int,Int), x,y)
Cosine(x)  = ccall((:Cosine, called), Float64, (Float64,), x)
function Sort(vals)
    ccall((:Sort, called), Void, (GoSlice,), GoSlice(vals))
    return vals     # agar sesuai standard
end
Hello(msg) = ccall((:Log, called), Int, (GoStr,), GoStr(msg))

##### Pemanggilan fungsi #####
for ex in [:(Add(93, 42)), :(Cosine(1)), :(Sort([39, 15, 9, 135, 27, 34]))]
    println("$ex = $(eval(ex))")
end
Hello("Xathrya")

File: Languages/C++/Go/loadlib/invoke.cpp
/**
Memanggil kode Go dari C++

Diuji pada:
    - Go 1.11.2 windows/amd64 + MinGW GCC 7.3.0

Compile:
    (GCC)
    % g++ invoke.cpp -o invoke 
*/

#include <iostream>
#include <windows.h>
#include "go-types.h"

/*
Beberapa tipe di Go dapat dipetakan secara langsung sebagai tipe primitif di C++, 
sebagian lagi dapat dianggap seagai tipe majemuk / struktur.

Umumnya ketika membuat DLL dengan Go, compiler akan membuat sebuah file header
untuk memudahkan interfacing. Namun dalam kasus tertentu kita hanya mendapatkan file
DLL tanpa adanya header. Untuk melakukan interfacing, kita perlu mendefinisikan sendiri
fungsi-fungsi yang diekspor oleh DLL.

File "go-types.h" merupakan versi strip dari setiap header yang dibentuk oleh compiler Go.
File ini berisi tipe-tipe primitif yang ada di Go sehingga interfacing dapat dilakukan.

*/

/** Daftar fungsi yang diekspor oleh DLL **/
typedef GoInt     (__stdcall* AddFunc)   (GoInt p0, GoInt p1);
typedef GoFloat64 (__stdcall* CosineFunc)(GoFloat64 p0);
typedef void      (__stdcall* SortFunc)  (GoSlice p0);
typedef void      (__stdcall* HelloFunc) (GoString p0);


int main()
{
    int result;

    // Memuat modul DLL ================================================
    // Sebelum mendapatkan alamat dari fungsi terlebih dahulu kita harus memuat pustaka
	// ke memory dan memetakannya ke area yang masih tersedia.
	// Kita dapat menyerahkan tanggung jawab ini ke API
	auto called = LoadLibrary("called.dll");
	if (called == nullptr) 
	{
		MessageBox(nullptr, "Cannot locate the DLL file", "Error", MB_OK);
		return 1;
	}
	
	// Cari fungsi berdasarkan nama ====================================
	AddFunc     Add    = (AddFunc)     GetProcAddress(called, "Add");
    CosineFunc  Cosine = (CosineFunc)  GetProcAddress(called, "Cosine");
    SortFunc    Sort   = (SortFunc)    GetProcAddress(called, "Sort");
    HelloFunc   Hello  = (HelloFunc)   GetProcAddress(called, "Hello");

	// Panggil fungsi ==================================================
	
    /*
    Example 1: fungsi yang membutuhkan 2 integer dan 
        mengembalikan sebuah integer
    */
    GoInt a = 93;
    GoInt b = 42;
    std::cout << "Add(a, b) is: " << Add(a, b) << std::endl;

    /*
    Example 2: fungsi yang membutuhkan 1 float dan mengembalikan float
    */
    std::cout << "Cosine(1.0) is: " << Cosine(1.0) << std::endl;

    /*
    Example 3: fungsi yang membutuhkan slice array
    */
    GoInt data[6] = { 39, 15, 9, 135, 27, 34 };
    GoSlice nums = { data, 6, 6 };
    Sort(nums);
    for (int i = 0; i < 6; i++)
        std::cout << "data[" << i << "] = " << data[i] << std::endl;

    /*
    Example 4: fungsi yang membutuhkan String
    */
    GoString msg = { "Xathrya", 7 };
    Hello(msg);

	// Bebaskan area memory jika modul DLL tak terpakai lagi.	
	FreeLibrary(called);
}